malariagen_data/anoph/hap_freq.py

```diff
@@ -0,0 +1,391 @@
+        chunks: base_params.chunks = base_params.native_chunks,
+        inline_array: base_params.inline_array = base_params.inline_array_default,
+    ) -> xr.Dataset:
+        # Load sample metadata.
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets, sample_query=sample_query
+        )
+
+        # Prepare sample metadata for cohort grouping.
+        df_samples = _prep_samples_for_cohort_grouping(
+            df_samples=df_samples,
+            area_by=area_by,
+            period_by=period_by,
+        )
+
+        # Group samples to make cohorts.
+        group_samples_by_cohort = df_samples.groupby(["taxon", "area", "period"])
+
+        # Build cohorts dataframe.
+        df_cohorts = _build_cohorts_from_sample_grouping(
+            group_samples_by_cohort=group_samples_by_cohort,
+            min_cohort_size=min_cohort_size,
+        )
+
+        # Early check for no cohorts.
+        if len(df_cohorts) == 0:
+            raise ValueError(
+                "No cohorts available for the given sample selection parameters and minimum cohort size."
+            )
+
+        # Access haplotypes.
+        ds_haps = self.haplotypes(
+            region=region,
+            sample_sets=sample_sets,
+            sample_query=sample_query,
+            chunks=chunks,
+            inline_array=inline_array,
+        )
+
+        # Early check for no SNPs.
+        if ds_haps.sizes["variants"] == 0:  # pragma: no cover
+            raise ValueError("No SNPs available for the given region.")
+
+        # Access genotypes.
+        gt = allel.GenotypeDaskArray(ds_haps["call_genotype"].data)
+        with self._dask_progress(desc="Compute haplotypes"):
+            gt = gt.compute()
+
+        # List all haplotypes
+        gt_hap = gt.to_haplotypes()
+        f_all, _, _ = haplotype_frequencies(gt_hap)
+
+        # Count haplotypes.
+        hap_freq: dict[np.int64, float] = dict()
+        hap_count: dict[np.int64, int] = dict()
+        hap_nob: dict[np.int64, int] = dict()
+        freq_cols = dict()
+        count_cols = dict()
+        nobs_cols = dict()
+        cohorts_iterator = self._progress(
+            df_cohorts.itertuples(), desc="Compute allele frequencies"
+        )
+        for cohort in cohorts_iterator:
+            cohort_key = cohort.taxon, cohort.area, cohort.period
+            cohort_key_str = cohort.taxon + "_" + cohort.area + "_" + str(cohort.period)
+            # We reset all frequencies, counts to 0 for each cohort, nobs is set to the number of haplotypes
+            n_samples = cohort.size
+            hap_freq = {k: 0 for k in f_all.keys()}
+            hap_count = {k: 0 for k in f_all.keys()}
+            hap_nob = {k: 2 * n_samples for k in f_all.keys()}
+            assert n_samples >= min_cohort_size
+            sample_indices = group_samples_by_cohort.indices[cohort_key]
+            loc_coh = [i in sample_indices for i in range(0, gt.shape[1])]
```

Comment:
Possibly simplification, if you have the sample indices you could use those directly via `take()`, rather than building a boolean mask and using `compress()`.